tests/utilities.py

```diff
@@ -7,3 +11,26 @@ def latest_request(mocked, request_type, path):
 
 def json_of_latest_request(r):
     return r[-1].kwargs["json"]
+
+
+def fingerprint_consistency_test(python_script: Text, tmp_path: Path):
```

Comment:
Is it possible to do this by passing in a function instead? And then calling that in different subprocesses? Would mean we could avoid the python code in strings which totally side-steps all our CI checks.

Reply:
Seems this might be possible using the `dill` module: https://stackoverflow.com/a/44206921

Another way without external dependency might be writing the function and then limiting the code string to importing and executing that function.

Reply:
what about using `multiprocessing` - would that expose the same hash problem?

Reply:
Will test! good idea

Reply:
Nope, multiprocessing doesn't work:

```python
>>> import multiprocessing
>>> hash("5df")
-1792932716691811005
>>> def hash_func():
>>>    print(hash("5df"))
    
>>> processes = [multiprocessing.Process(target=hash_func) for i in range(3)]
>>> for p in processes:
>>>    p.start()
    
-1792932716691811005
-1792932716691811005
-1792932716691811005
```

Reply:
ah damn that's annoying. Ok well I reckon we can just remove these tests then. I think it's really good we confirmed that it works across runs, but code in a string like this is not maintainable in the long run. 

Once we've got all the pieces in place we can have an integration level test which checks that the graph doesn't need to re-run if nothing changes which will in turn check all the fingerprints are stable.

Reply:
Alrighty, will remove the tests then!

Reply:
Although it will make it more difficult to catch it if someone ever uses a hash function that isn't consistent across runs. Also makes it more difficult to evaluate different hashes potentially. Like while working on this I've came across the `xxhash` hash library which is apparently much faster than md5. https://github.com/Cyan4973/xxHash

Reply:
It should be covered by a full integration test which does a whole train and another without changing training data. If any of the fingerprints aren't stable then nodes will re-run. Only if all are stable will the whole graph be cached.